gpytorch/__init__.py

```diff
@@ -1,37 +1,250 @@
+
+    Returns:
+        - matrix nxk - (mat1)^{-1}mat2
+    """
     return Invmm()(mat1, mat2)
 
 
 def invmv(mat, vec):
+    """
+    Computes a linear solve with a single right hand side
+
+    Args:
+        - mat1 (matrix nxn) - Matrix to solve with
+        - vec (vector n) - Right hand side vector
+
+    Returns:
+        - vector n - (mat1)^{-1}vec
+    """
     return Invmv()(mat, vec)
 
 
 def normal_cdf(x):
+    """
+    Computes the element-wise standard normal CDF of an input tensor x.
+    """
     return NormalCDF()(x)
 
 
 def log_normal_cdf(x):
+    """
+    Computes the element-wise log standard normal CDF of an input tensor x.
+
+    This function should always be preferred over calling normal_cdf and taking the log
+    manually, as it is more numerically stable.
+    """
     return LogNormalCDF()(x)
 
 
 def mvn_kl_divergence(mean_1, chol_covar_1, mean_2, covar_2):
+    """
+    Computes the KL divergence between two Gaussian distributions N1 and N2.
+
+    N1 is parameterized by a mean vector and the Cholesky decomposition of the covariance matrix.
+    N2 is parameterized by a mean vector and covariance matrix.
+
+    For more information, see the MVNKLDivergence function documentation.
+    """
     return MVNKLDivergence()(mean_1, chol_covar_1, mean_2, covar_2)
+
+
+def interpolated_toeplitz_mul(c, W_left, W_right, v, noise_diag=None):
+    """
+    Given a interpolated symmetric Toeplitz matrix W_left*T*W_right, plus possibly an additional
+    diagonal component s*I, compute a matrix-vector product with some vector or matrix v.
+
+    Args:
+        - c (vector m) - First column of the symmetric Toeplitz matrix T
+        - W_left (sparse matrix nxm) - Left interpolation matrix
+        - W_right (sparse matrix pxm) - Right interpolation matrix
+        - v (matrix pxk) - Vector (k=1) or matrix (k>1) to multiply WTW with
+        - noise_diag (vector p) - If not none, add (s*I)v to WTW at the end.
+
+    Returns:
+        - matrix nxk - The result of multiplying (WTW + sI)v if noise_diag exists, or (WTW)v otherwise.
+    """
+    noise_term = None
+    if v.ndimension() == 1:
+        if noise_diag is not None:
+            noise_term = noise_diag.expand_as(v) * v
+        v = v.unsqueeze(1)
+        mul_func = ToeplitzMV()
+    else:
+        if noise_diag is not None:
+            noise_term = noise_diag.unsqueeze(1).expand_as(v) * v
+        mul_func = ToeplitzMM()
+
+    # Get W_{r}^{T}v
+    Wt_times_v = torch.dsmm(W_right.t(), v)
+    # Get (TW_{r}^{T})v
+    TWt_v = mul_func.forward(c, c, Wt_times_v.squeeze())
+
+    if TWt_v.ndimension() == 1:
+        TWt_v.unsqueeze_(1)
+
+    # Get (W_{l}TW_{r}^{T})v
+    WTWt_v = torch.dsmm(W_left, TWt_v).squeeze()
+
+    if noise_term is not None:
+        # Get (W_{l}TW_{r}^{T} + \sigma^{2}I)v
+        WTWt_v = WTWt_v + noise_term
+
+    return WTWt_v
+
+
+def _exact_predict(test_mean, test_test_covar, train_y, train_mean,
```

Comment:
It would be good to improve this, so that all the branching logic is handled in separate classes.